text/0000-with-bounds-in-pi-types.md

```diff
@@ -0,0 +1,685 @@
+
+This RFC introduces `with` bounds on type level values, making it possible to
+set constraints on the constant parameters introduced in implementations,
+functions, structs, and other items.
+
+# Motivation
+[motivation]: #motivation
+
+There are many usecases. These allows certain often requested features to live
+in standalone libraries (e.g.,
+[bounded-integers](#bounded-integersinterval-arithmetics), [type level
+numerals](#array-generics), [statically checked
+indexing](#statically-checked-indexing), lattice types).
+
+I've listed examples of such usecases and their implementations below.
+
+## Examples
+
+### Bounded integers/interval arithmetics
+
+One can define the so called "bounded integers" (integers which carry an
+upper and lower bound, checked at compile time):
+
+```rust
+use std::ops;
+
+/// A bounded integer.
+///
+/// This has two value parameter, respectively representing an upper and a lower bound.
+pub struct BoundedInt<const lower: usize, const upper: usize> {
+    /// The inner runtime value.
+    ///
+    /// Note how this is not public. This is because we cannot, in the updated
+    /// version of the RFC, set constraints on compile time, so when created it
+    /// must either be dynamically ensured or an unsafe contract must be made
+    in order to satisfy the invariants.
+    n: usize,
+}
+
+// To see how this holds the `with` clause above, see the section on `identities`.
+impl<const n: usize> BoundedInt<n, n> {
+    fn new() -> Self {
+        BoundedInt {
+            n: n,
+        }
+    }
+}
+
+/// Addition of two `BoundedInt` will simply add their bounds.
+///
+/// We check for overflow making it statically overflow-free calculations.
+impl<const upper_a: usize,
+     const lower_a: usize,
+     const upper_b: usize,
+     const lower_b: usize> ops::Add<BoundedInt<lower_b, upper_b>> for BoundedInt<lower_a, upper_a>
+     // We have to satisfy the constraints set out in the struct definition.
+     with lower_a <= upper_a,
+           lower_b <= upper_b,
+           // Check for overflow by some `const fn`.
+           is_overflow_safe(upper_a, upper_b) {
+    // These parameters are constant expression.
+    type Output = BoundedInt<lower_a + lower_b, upper_a + upper_b>;
+
+    fn add(self, rhs: BoundedInt<lower_b, upper_b>) -> Self::Output {
+        BoundedInt {
+            n: self.n + rhs.n,
+        }
+    }
+}
+
+impl<const upper_a: usize,
+     const lower_a: usize,
+     const upper_b: usize,
+     const lower_b: usize> From<BoundedInt<lower_b, upper_b>> for BoundedInt<lower_a, upper_a>
+     with lower_a <= upper_a,
+           lower_b <= upper_b,
+           // We will only extend the bound, never shrink it without runtime
+           // checks, thus we add this clause:
+           lower_b <= lower_a && upper_b >= upper_a {
+    fn from(from: BoundedInt<lower_b, upper_b>) -> Self {
+        BoundedInt {
+            n: from.n,
+        }
+    }
+}
```

Comment:
if `lower_b <= lower_a && upper_b >= upper_a` hold true, then [lower_b, upper_b] actually contains [lower_a, upper_a]. So if you cast one number from `BoundedInt<lower_b, upper_b>` to `BoundedInt<lower_a, upper_a>`, you actually "shrink" the bound rather than extend it.